### DISCUSSION TITLE
Alternate discovery method for localhost OPC Servers?

### DISCUSSION THREAD

**Discussion by seversonic** [FreeOpcUa/opcua-asyncio#1556] [Q&A] [answered]
I wrote a script using this library that initializes an OPC-UA server after it's first install. 

This script works when connecting to an OPC server that is on the network, but not when trying to connect to localhost. 

UAExpert detects the server (Beckhoff's TF6100), but UADiscover does not. 

![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/43625790/1c057740-761b-4a84-9d7b-d007b923254c)

The server discovered on left side of the image in UAExpert is from Beckhoff
![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/43625790/77ad4f84-2433-4092-8272-09931e77f797)

On the right, the first `uadiscover` call is made when there are no other OPC Servers running, I then started the uaserver tool just to prove that the tool is working (it is). 

What information do I need from the vendor to use this library to discovery the server locally?

The only error I get back is shown as `[Errno 111] Connect call failed` , a description that I could not find in this repo. 

Thanks!

**Comment by schroeder-** [comment]
Could there be that the server isn't binding on 127.0.0.1? Maybe try to get your local ip address and connect to it. Or try using hostname:
```
host = socket.gethostbyname("localhost")
url = f"opc.tcp://{host}:4840"
```

**Reply by seversonic** [reply]
I did try with hostname, statically writing 127.0.0.1 and every other variation I could think of. 

I also tried your code, but still got the same error. 

I enabled debug logging, but it doesnt show much more. 
![image](https://github.com/FreeOpcUa/opcua-asyncio/assets/43625790/3a6a09d1-f304-4a68-849d-f9e1b45afbbe)

**Reply by schroeder-** [reply]
Rly strange. We just use sockets from asyncio. There must be something special how the local server binds...

**Reply by seversonic** [reply]
I have a contact at Beckhoff who I could reach out to, he has some software experience within the company. 

I just dont know what information to ask them to find out how to make this connect?

**Reply by seversonic** [reply]
I am running this on WSL, if that complicates things

**Reply by schroeder-** [reply]
This should apply to your problem: https://stackoverflow.com/questions/68602075/python-sqli-scripts-on-wsl2-errno-111

**Reply by seversonic** [reply, marked as answer]
Interesting!

I moved to a windows installation of python and this library and it detects the localhost now when I run UAdiscover.exe

Now I just have to resolve the issue with setup_self_signed_certificate adding extra \\ in the path.
